### PR TITLE
Jenkins CI/CD & Openshift Bootstrap

### PR DESCRIPTION
**Pull request by jujaga**
# Description

This PR adds Jenkins tooling and Openshift templates with the aim of making the application deployable via CI/CD.

- https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-117

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation (non-breaking change with enhancements to documentation)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

This PR is a slight leap of faith as that the code must be in master before it can be executed upon.

### CODE REVIEW

File: tools/jenkins/README.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+### create config map for related namespaces
+
+```
+oc -n z208i4-tools process -f 'openshift/ns-config.json' -p 'DEV=z208i4-dev' -p 'TEST=z208i4-test' -p 'PROD=z208i4-prod' -p 'TOOLS=z208i4-tools' | oc  -n z208i4-tools create -f -
+```
+
+### create config map for the application
+
+```
+oc -n z208i4-tools process -f 'openshift/jobs-config.json' -p 'REPO_OWNER=bcgov' -p 'REPO_NAME=nr-get-token' -p 'APP_NAME=getok' -p 'APP_DOMAIN=pathfinder.gov.bc.ca' | oc -n z208i4-tools create -f -
```
Comment (loneil) [inline tools/jenkins/README.md]: getok reference

File: tools/jenkins/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+```
+
+
+### process the build config templates...
+
+These build configs have no build triggers, we start them manually (or in Jenkins job script)
+#### master
+
+```
+oc -n z208i4-tools process -f 'openshift/build-master.yaml' -p 'NAME=jenkins' -p 'SUFFIX=-prod' -p 'VERSION=prod-1.0.0' -p 'SOURCE_REPOSITORY_URL=https://github.com/bcgov/nr-get-token' -p 'SOURCE_REPOSITORY_REF=master' -l app-name=jenkins -l env-name=prod -l env-id=0 -l github-repo=https://github.com/bcgov/nr-get-token -l github-owner=bcgov -l app=jenkins-prod -o yaml | oc -n z208i4-tools create -f -
```
Comment (loneil) [inline tools/jenkins/README.md]: nr-get-token references

File: tools/jenkins/README.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+##### build and follow...
+
+```
+oc -n z208i4-tools start-build bc/jenkins-prod -F
+```
+
+#### slave
+
+```
+oc -n z208i4-tools process -f 'openshift/build-slave.yaml' -p 'NAME=jenkins' -p 'SUFFIX=-prod' -p 'VERSION=prod-1.0.0' -p 'SLAVE_NAME=main' -p 'SOURCE_IMAGE_STREAM_TAG=jenkins:prod-1.0.0' -l app-name=jenkins -l env-name=prod -l env-id=0 -l github-repo=https://github.com/bcgov/nr-get-token -l github-owner=bcgov -l app=jenkins-prod -o yaml | oc -n z208i4-tools create -f -
```
Comment (loneil) [inline tools/jenkins/README.md]: nr-get-token reference

File: tools/jenkins/README.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+oc -n z208i4-tools start-build bc/jenkins-slave-main-prod -F
+```
+
+
+### process the deployment templates
+
+#### master
+
+```
+oc -n z208i4-tools process -f 'openshift/deploy-master.yaml' -p 'NAME=jenkins' -p 'SUFFIX=-prod' -p 'VERSION=prod-1.0.0' -p 'ROUTE_HOST=jenkins-prod-z208i4-tools.pathfinder.gov.bc.ca' -p 'GH_USERNAME=bcgov-nr-csst' -p 'GH_PASSWORD=<personal_access_token>' -l app-name=jenkins -l env-name=prod -l env-id=0 -l github-repo=https://github.com/bcgov/nr-get-token -l github-owner=bcgov -l app=jenkins-prod -o yaml | oc -n z208i4-tools create -f -
```
Comment (loneil) [inline tools/jenkins/README.md]: nr-get-token reference

File: tools/jenkins/README.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+```
+oc -n z208i4-dev policy add-role-to-user 'admin' 'system:serviceaccount:z208i4-tools:jenkins-prod'
+oc -n z208i4-test policy add-role-to-user 'admin' 'system:serviceaccount:z208i4-tools:jenkins-prod'
+oc -n z208i4-prod policy add-role-to-user 'admin' 'system:serviceaccount:z208i4-tools:jenkins-prod'
+```
+
+#### slave
+
+```
+oc -n z208i4-tools process -f 'openshift/deploy-slave.yaml' -p 'NAME=jenkins' -p 'SUFFIX=-prod' -p 'VERSION=prod-1.0.0' -p 'SLAVE_NAME=build' -p 'SLAVE_LABELS=build deploy test ui-test' -p 'SLAVE_EXECUTORS=3' -p 'CPU_REQUEST=300m' -p 'CPU_LIMIT=500m' -p 'MEMORY_REQUEST=2Gi' -p 'MEMORY_LIMIT=2Gi' -l app-name=jenkins -l env-name=prod -l env-id=0 -l github-repo=https://github.com/bcgov/nr-get-token -l github-owner=bcgov -l app=jenkins-prod -o yaml | oc -n z208i4-tools create -f -
```
Comment (loneil) [inline tools/jenkins/README.md]: nr-get-token reference

File: tools/jenkins/README.md (hunk 1 above)
Comment (jujaga) [reply tools/jenkins/README.md]: Fixed

File: tools/jenkins/README.md (hunk 2 above)
Comment (jujaga) [reply tools/jenkins/README.md]: Fixed

File: tools/jenkins/README.md (hunk 4 above)
Comment (jujaga) [reply tools/jenkins/README.md]: Fixed

File: tools/jenkins/README.md (hunk 5 above)
Comment (jujaga) [reply tools/jenkins/README.md]: Fixed

File: tools/jenkins/README.md (hunk 3 above)
Comment (jujaga) [reply tools/jenkins/README.md]: Fixed

**Pull request merged**